### PR TITLE
Rustdoc: use ThinVec for GenericArgs bindings

### PR DESCRIPTION
**Pull request by Kobzol**
The bindings are almost always empty. This reduces the size of `PathSegment` and `GenericArgs` by about one fourth.

### CODE REVIEW

**Comment by Kobzol** [comment]
FWIW, I also tried to go through all Rustdoc `clean` structures and reduce their sizes by converting all `Vec<T>` to `Box<[T]>`, which reduces `24` bytes to `16` bytes.

Sadly, it seems that it increased RSS instead of reducing it, probably because in most cases the Vec still has to be created first, and also because `into_boxed_slice` can reallocate, which is unfortunate. Something like https://docs.rs/thin-slice/latest/thin_slice/ could be used to reduce the size further to `8` bytes, but I'm afraid that the problems with boxing the slice will prevail.

**Comment by cjgillot** [comment]
r? rust-lang/rustdoc

File: compiler/rustc_data_structures/src/thin_vec.rs [hunk 1]
```diff
@@ -121,6 +123,20 @@ impl<T: HashStable<CTX>, CTX> HashStable<CTX> for ThinVec<T> {
...
+    }
+}
+
+impl<T: PartialEq> PartialEq for ThinVec<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.deref().eq(other.deref())
+    }
+}
+
+impl<T: PartialEq> Eq for ThinVec<T> {}
```
Comment (camelid) [inline compiler/rustc_data_structures/src/thin_vec.rs:138]: Why not just `derive(Hash, PartialEq, Eq)`?

File: src/librustdoc/clean/auto_trait.rs [hunk 2]
```diff
@@ -568,10 +568,10 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
                                 // Convert something like '<T as Iterator::Item> = u8'
                                 // to 'T: Iterator<Item=u8>'
                                 GenericArgs::AngleBracketed { ref mut bindings, .. } => {
-                                    bindings.push(TypeBinding {
+                                    bindings.extend([TypeBinding {
```
Comment (camelid) [inline src/librustdoc/clean/auto_trait.rs:571]: Why this change?

**Review by camelid — CHANGES_REQUESTED** [review]

File: compiler/rustc_data_structures/src/thin_vec.rs (hunk 1 above)
Comment (Kobzol) [reply compiler/rustc_data_structures/src/thin_vec.rs:138]: It didn't compile with Derive, but I suspect it was a fluke, I'll try again.

File: src/librustdoc/clean/auto_trait.rs (hunk 2 above)
Comment (Kobzol) [reply src/librustdoc/clean/auto_trait.rs:571]: ThinVec doesn't provide a push method. Should I add it?

Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs:571]: Yes, please.

**Comment by camelid** [comment]
Small instructions count improvements and mostly moderate improvements on max-rss (with a few regressions, but perhaps spurious).

File: compiler/rustc_data_structures/src/thin_vec.rs [hunk 3]
```diff
@@ -1,11 +1,12 @@
 use crate::stable_hasher::{HashStable, StableHasher};
+use std::hash::Hash;
```
Comment (camelid) [inline compiler/rustc_data_structures/src/thin_vec.rs:2]: This import is unnecessary.

File: compiler/rustc_data_structures/src/thin_vec.rs [hunk 4]
```diff
@@ -20,6 +20,13 @@ impl<T> ThinVec<T> {
     pub fn iter_mut(&mut self) -> std::slice::IterMut<'_, T> {
         self.into_iter()
     }
+
+    pub fn push(&mut self, item: T) {
```
Comment (camelid) [inline compiler/rustc_data_structures/src/thin_vec.rs:24]: Hmm, in an old version of this PR, you used `.extend([...])` with one element, so I requested adding `fn push()`. But now that it's added, I don't see any (new) uses of either `extend` or `push` in rustdoc. Do we still need `push`? Am I missing something?

Comment (Kobzol) [reply compiler/rustc_data_structures/src/thin_vec.rs:24]: The old code called push on a Vec. I added a ThinVec, but it didn't have push, so I changed the push calls to extend.

But then I added the push method to ThinVec, so I reverted the method call changes. There is no diff for these now, since the code is the same - push is called.

Comment (camelid) [reply compiler/rustc_data_structures/src/thin_vec.rs:24]: Ah, of course! Thanks :)

**Review by camelid — APPROVED** [review]
Thanks!
